0000-overhaul-qiskit-experiments.md

```diff
@@ -0,0 +1,638 @@
+    for q in range(backend.num_qubits)
+  ],
+  backend=backend,
+)
+exp_data = exp.run()
+```
+
+A component experiment (`T1`) creates M circuit, and the parallel experiment is defined for N qubits.
+In principle number of task for this experiment is M*N, but the parallel experiment digests N circuits of the components and
+creates a single merged circuit to run. Thus number of actual payload is always M, regardless of qubit number.
+It seems like QE framework is syntactically scalable, however, data analysis overhead will almost monotonically increase with N in the current implementation.
+
+Although circuit generation cost and communication overhead are also significant, we should focus on analysis task in this RFC document.
+This is because circuit generation heavily relies on the core [Qiskit](https://github.com/Qiskit/qiskit-terra) package, 
+and communication overhead depends on the payload format that [Provider](https://github.com/Qiskit/qiskit-ibm-provider) defines (currently IBM uses QPY binary format).
+Especially, we put attention in the [curve analysis](https://github.com/Qiskit-Extensions/qiskit-experiments/tree/main/qiskit_experiments/curve_analysis) module,
+because almost all experiments in the maintenance workflow require parameter scan and fitting to find an optimal control parameter or device parameter as shown in above example.
+
+## Motivation
+
+The plot below shows wall clock time (with and without plotting the Matplotlib figure) for running parallel T1 analysis against number of qubits on my laptop; MBP Intel Core i7 @ 2.3 GHz (4 cores), 32 GB Mem.
+As shown in this experiment result, required time increases roughly quadratically with qubit number.
+To run a single experiment on few hundreds qubit device (such as IBM Osprey processor), it would take 10s minutes to get analysis results.
+This is non-negligible downtime for a quantum computing system because the system calibration workflow often consists of [dependent tasks](https://arxiv.org/abs/1803.03226),
+and the following experiment run might be on hold until predecessor results become available.
+
+<img src="./0000-overhaul-qiskit-experiments/scaling.png" style="width: 6in">
+
+(see [this gist](https://gist.github.com/nkanazawa1989/41428cd21b7307a70be720779b364ca3) for more details.)
+
+I also scrutinized the performance of the current framework with a statistical profiler.
+Since the percentage of function detection varied significantly from run to run, I only show qualitative results here.
+I found following subroutines tend to frequently show up in the frame graph.
+
+#### Initialization of `ExperimentData` sub-container
+
+Because composite analysis needs to instantiate new data container to run component analyses,
+calling the constructor of this expensive class consumes significant time in heavily batched experiment.
+Especially, acquisition of a service object and qiskit version information is the expensive part.
```

Comment:
Should we recommend working in a Python prompt for some use-cases, such that this overhead is paid once?

Reply:
Can you elaborate? I am thinking to introduce simpler container that doesn't have connection to experiment service. I think current design of `_run_analysis(self, experiment_data)` gives too much freedom to experiment developers, since technically they can modify remote data during analysis. This could introduce some edge case that we don't want to support.

Reply:
I was referring to caching service object and qiskit version in the a python session (say IPython). A container without experiments service is obviously useful for certain use-cases, but what about those use-cases that a user will require results DB access?

Reply:
I meant analysis usually doesn't require results DB access, so we can introduce simplified internal data container that doesn't have experiment service. Caching service object is easy and effective, but do we really need service object to run experiment analysis?